theano/gof/utils.py

```diff
@@ -217,14 +217,21 @@ def __call__(cls, *args, **kwargs):
         props = cls.__dict__.get('__props__', None)
         if props is None:
             return super(MetaObject, cls).__call__(*args, **kwargs)
+
         key = [cls]
         for arg in args:
             arg = make_hashable(arg)
             key.append(arg)
+            arg_dtype = get_arg_dtype(arg)
+            if arg_dtype:
+                key.extend(arg_dtype)
         for k, v in kwargs.items():
             key.append(k)
             v = make_hashable(v)
             key.append(v)
+            v_dtype = get_arg_dtype(v)
+            if v_dtype:
+                key.extend(v_dtype)
```

Comment:
@ReyhaneAskari Why this? This remove the value from the key. The value is important.
@Sentient07 your comment with the one from @ReyhaneAskari about caching too much stuff make me think of something. If one of the kwargs key isn't a prop, we could skip the cache. This with the elemwise name parameter that isn't a props, would allow the python gc to do much of its work.

@ReyhaneAskari can you try that? If that isn't enough, we could try to just skip the cache for all elemwise.